src/status.ts

```diff
@@ -2,258 +2,385 @@ import * as vscode from 'vscode';
+    return this._getCurrentText();
+  }
+  isVisible():boolean {
+    return this._isBusy || true;
+  }
+  protected getType():ButtonType {
+    return this._config.statusbar.build.type;
+  }
+}
+
+export class StatusBar implements vscode.Disposable {
+  private readonly _kitSelectionButton = new KitSelection(this._config, 3.6);
+  private readonly _cmakeToolsStatusItem = new CMakeStatus(this._config, 3.55);
 
-  showWarningMessage(msg: string) {
-    this._warningMessage.color = 'yellow';
-    this._warningMessage.text = `$(alert) ${msg}`;
-    this._warningMessage.show();
-    setTimeout(() => this._warningMessage.hide(), 5000);
+  private readonly _activeFolderButton = new ActiveFolderButton(this._config, 3.5);
+  private readonly _buildTargetNameButton = new BuildTargetSelectionButton(this._config, 3.45);
+  private readonly _buildButton:BuildButton = new BuildButton(this._config, 3.4);
+
+  private readonly _launchTargetNameButton = new LaunchTargetSelectionButton(this._config, 3.35);
+  private readonly _debugButton:DebugButton = new DebugButton(this._config, 3.3);
+  private readonly _runButton = new LaunchButton(this._config, 3.25);
+
+  private readonly _testButton = new CTestButton(this._config, 3.2);
+
+  private readonly _buttons: Button[];
+
+  constructor(private readonly _config: ConfigurationReader) {
+    this._buttons = [
+      this._activeFolderButton,
+      this._cmakeToolsStatusItem,
+      this._kitSelectionButton,
+      this._buildTargetNameButton,
+      this._launchTargetNameButton,
+      this._debugButton,
+      this._buildButton,
+      this._testButton,
+      this._runButton
+    ];
+    this._config.onChange('statusbar', ()=>this.update());
+    this.update();
   }
 
-  private _hideDebugButton = false;
+  dispose = () => this._buttons.forEach(btn => btn.dispose());
+  update = () => this._buttons.forEach(btn => btn.update());
+  setVisible= (v: boolean) => this._buttons.forEach(btn => btn.forceHidden = !v);
+
+  setActiveFolderName = (v: string) => this._activeFolderButton.text = v;
+  setAutoSelectActiveFolder = (autoSelectActiveFolder: boolean) => this._activeFolderButton.autoSelect = autoSelectActiveFolder;
+  setBuildTypeLabel = (v: string) => this._cmakeToolsStatusItem.buildTypeLabel = v;
+  setStatusMessage = (v: string) => this._cmakeToolsStatusItem.statusMessage = v;
+  setBuildTargetName = (v: string) => this._buildTargetNameButton.text = v;
+  setLaunchTargetName = (v: string) => this._launchTargetNameButton.text = v;
+  setCTestEnabled = (v: boolean) => this._testButton.enabled = v;
+  setTestResults = (v: BasicTestResults|null) => this._testButton.results = v;
+  setIsBusy = (v:boolean) => this._buildButton.isBusy = v;
+  setActiveKitName = (v:string) => this._kitSelectionButton.text = v;
+
   hideDebugButton(shouldHide: boolean = true) {
-    this._hideDebugButton = shouldHide;
-    this.reloadVisibility();
+    if (false) return shouldHide;
+    // TODO: Find out if this is called...
```

Comment:
This is a feature for other extensions that allows them to configure our behavior a bit. You can remove the comment.

Reply:
ok.
i added this to the button
same with hideLaunchButton.